File: src/cs/IdpProtocol/MasterNode.cs
// Copyright (c) VitalElement. All rights reserved.
// Licensed under the MIT license. See licence.md file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace IdpProtocol
{
    public enum NodeEnumerationState
    {
        Pending,
        DetectingRouter,
        EnumeratingNodes,
        StartEnumeratingAdaptors,
        EnumeratingAdaptors,
        Idle
    };

    public class NodeInfo
    {
        public NodeInfo(NodeInfo parent, UInt16 address)
        {
            Parent = parent;
            Address = address;
            LastSeen = DateTime.Now;
            Guid = Guid.Empty;
            EnumerationState = NodeEnumerationState.Idle;
            Parent = parent;
            Name = null;
            Children = new List<NodeInfo>();
        }

        public UInt16 Address { get; }

        internal DateTime LastSeen { get; set; }

        public Guid Guid { get; internal set; }

        public NodeInfo Parent { get; }

        public string Name { get; internal set; }

        internal NodeEnumerationState EnumerationState { get; set; }

        public List<NodeInfo> Children { get; }

        public bool IsRouter => Guid == IdpRouter.RouterGuid;
    }

    public class MasterNode : IdpNode
    {
        private UInt16 _nextAddress;
        private Stack<UInt16> _freeAddresses;
        private Dictionary<UInt16, NodeInfo> _nodeInfo;
        private NodeInfo _root;
        NodeInfo _currentEnumerationNode;

        public const UInt16 MasterNodeAddress = 0x0001;
        public static readonly Guid MasterGuid = Guid.Parse("554C0A67-F228-47B5-8155-8C5436D533DA");

        public MasterNode() : base(MasterGuid, "Network.Master", MasterNodeAddress)
        {
            _freeAddresses = new Stack<ushort>();
            _nodeInfo = new Dictionary<ushort, NodeInfo>();

            _nextAddress = 2;
            NodeTimeout = 5000;
            _root = new NodeInfo(null, MasterNodeAddress);
            _root.Guid = MasterGuid;

            _root.Name = "Network.Master";
            _root.EnumerationState = NodeEnumerationState.Pending;

            _currentEnumerationNode = null;

            _nodeInfo[Address] = _root;

            Manager.RegisterResponseHandler((UInt16)NodeCommand.Ping, response =>
            {
                OnPollResponse(response.Transaction.Source);
            });
        }

        public bool IsEnumerating { get; private set; }

        protected override void OnReset()
        {

        }

        private NodeInfo FindNode(UInt16 address)
        {
            if (_nodeInfo.ContainsKey(address))
            {
                return _nodeInfo[address];
            }

            return null;
        }

        private UInt16 GetFreeAddress()
        {
            if (_freeAddresses.Count == 0)
            {
                return _nextAddress++;
            }
            else
            {
                return _freeAddresses.Pop();
            }
        }

        private void OnPollResponse(UInt16 address)
        {
            if (address != Address)
            {
                var node = FindNode(address);

                if (node != null)
                {
                    node.LastSeen = DateTime.Now;
                }
            }
        }

        private void InvalidateNodes()
        {
            var currentTime = DateTime.Now;

            foreach (var node in _nodeInfo.Values.ToList())
            {
                if (node != _root && currentTime >= node.LastSeen + TimeSpan.FromMilliseconds(NodeTimeout))
                {
                    _nodeInfo.Remove(node.Address);

                    _freeAddresses.Push(node.Address);
                }
            }
        }

        private async Task OnNodeAdded(UInt16 parentAddress, UInt16 address)
        {
            _nodeInfo[address] = new NodeInfo(_nodeInfo[parentAddress], address);

            _nodeInfo[parentAddress].Children.Add(_nodeInfo[address]);

            var (success, response) = await SendRequestAsync(address, OutgoingTransaction.Create((UInt16)NodeCommand.GetNodeInfo, CreateTransactionId()));

            if (response != null && response.ResponseCode == IdpResponseCode.OK)
            {
                _nodeInfo[address].Guid = response.Transaction.Read<Guid>();

                _nodeInfo[address].Name = response.Transaction.ReadUtf8String();

                // TODO read string.

                if (_nodeInfo[address].IsRouter)
                {
                    _nodeInfo[address].EnumerationState = NodeEnumerationState.Pending;
                }
                else
                {
                    _nodeInfo[address].EnumerationState = NodeEnumerationState.Idle;
                }

                // normally on enumerate called here...
            }
        }

        private void VisitNodes(NodeInfo root, Func<NodeInfo, bool> visitor)
        {
            if (visitor(root))
            {
                foreach (var child in root.Children)
                {
                    VisitNodes(child, visitor);
                }
            }
        }

        private NodeInfo GetNextEnumerationNode()
        {
            NodeInfo result = null;

            VisitNodes(_root, node =>
            {
                if (node.EnumerationState != NodeEnumerationState.Idle)
                {
                    result = node;
                    return false;
                }

                return true;
            });

            return result;
        }

        private async Task EnumerateAsync()
        {
            NodeInfo node = null;

            while (true)
            {
                node = GetNextEnumerationNode();

                if (node == null)
                {
                    break;
                }

                _currentEnumerationNode = node;

                if (node == _root)
                {
                    _currentEnumerationNode.EnumerationState = NodeEnumerationState.DetectingRouter;

                    await DetectRouter();
                }
                else if (node.IsRouter)
                {
                    switch (node.EnumerationState)
                    {
                        case NodeEnumerationState.Pending:
                        case NodeEnumerationState.EnumeratingNodes:
                            await EnumerateRouterNode(node.Address);
                            break;

                        case NodeEnumerationState.StartEnumeratingAdaptors:
                            await StartEnumeratingAdaptors(node.Address);
                            break;

                        case NodeEnumerationState.EnumeratingAdaptors:
                            await EnumerateRouterAdaptor(node.Address);
                            break;

                        default:
                            break;
                    }
                }
            }
        }

        private async Task DetectRouter()
        {
            var address = GetFreeAddress();

            var (success, response) = await SendRequestAsync(0xFFFF, OutgoingTransaction.Create((UInt16)NodeCommand.RouterDetect, CreateTransactionId()).Write(address));

            if (success && response != null && response.ResponseCode == IdpResponseCode.OK)
            {
                var enumerated = response.Transaction.Read<bool>();

                if (enumerated)
                {
                    _currentEnumerationNode.EnumerationState = NodeEnumerationState.Idle;

                    await OnNodeAdded(Address, address);

                    return;
                }
            }

            _freeAddresses.Push(address);

            // normally on enumerate called.
        }

        private async Task EnumerateRouterNode(UInt16 routerAddress)
        {
            var address = GetFreeAddress();

            var (success, response) = await SendRequestAsync(routerAddress, OutgoingTransaction.Create((UInt16)NodeCommand.RouterEnumerateNode, CreateTransactionId()).Write(address));

            if (response != null && response.ResponseCode == IdpResponseCode.OK)
            {
                var enumerated = response.Transaction.Read<bool>();

                if (enumerated)
                {
                    await OnNodeAdded(routerAddress, address);

                    return;
                }
            }

            _freeAddresses.Push(address);

            _currentEnumerationNode.EnumerationState = NodeEnumerationState.StartEnumeratingAdaptors;
        }

        private async Task StartEnumeratingAdaptors(UInt16 routerAddress)
        {
            var (success, response) = await SendRequestAsync(routerAddress,
                                                  OutgoingTransaction.Create((UInt16)NodeCommand.RouterPrepareToEnumerateAdaptors, CreateTransactionId()));

            if (response != null && response.ResponseCode == IdpResponseCode.OK)
            {
                if (_currentEnumerationNode != null)
                {
                    _currentEnumerationNode.EnumerationState = NodeEnumerationState.EnumeratingAdaptors;
                }
            }
            else
            {
                if (_currentEnumerationNode != null)
                {
                    _currentEnumerationNode.EnumerationState = NodeEnumerationState.Idle;
                }
            }
        }

        private async Task EnumerateRouterAdaptor(UInt16 routerAddress)
        {
            var address = GetFreeAddress();

            var responseTask = Manager.WaitForResponseAsync((UInt16)NodeCommand.RouterDetect, 500);

            var (success, response) = await SendRequestAsync(routerAddress, OutgoingTransaction.Create((UInt16)NodeCommand.RouterEnumerateAdaptor, CreateTransactionId()).Write(address));

            if (response != null && response.ResponseCode == IdpResponseCode.OK)
            {
                var enumerated = response.Transaction.Read<bool>();

                if (enumerated)
                {
                    var routerDetectResponse = await responseTask;

                    if (routerDetectResponse != null && routerDetectResponse.ResponseCode == IdpResponseCode.OK)
                    {
                        var routerEnumerated = routerDetectResponse.Transaction.Read<bool>();

                        if (routerEnumerated)
                        {
                            await OnNodeAdded(routerAddress, address);

                            return;
                        }

                    }
                }
            }

            _freeAddresses.Push(address);

            _currentEnumerationNode.EnumerationState = NodeEnumerationState.Idle;
        }

        public bool HasNode(UInt16 address)
        {
            return _nodeInfo.ContainsKey(address);
        }

        public NodeInfo GetNodeInfo(UInt16 address)
        {
            return _nodeInfo[address];
        }

        public async Task EnumerateNetworkAsync()
        {
            if (Connected)
            {
                IsEnumerating = true;

                VisitNodes(_root, node =>
                {
                    if (node.IsRouter)
                    {
                        node.EnumerationState = NodeEnumerationState.Pending;
                    }
                    else if (node != _root)
                    {
                        node.EnumerationState = NodeEnumerationState.Idle;
                    }

                    return true;
                });

                await EnumerateAsync();

                IsEnumerating = false;
            }
        }

        public void PollNetwork()
        {
            //InvalidateNodes();

            SendRequest(0x0000, OutgoingTransaction.Create(0xA002, CreateTransactionId()));
        }

        public void ResetNetwork()
        {
            SendRequest(0x0000, OutgoingTransaction.Create((UInt16)NodeCommand.Reset, CreateTransactionId(), IdpCommandFlags.None));
        }

        public void TraceNetworkTree(NodeInfo node = null, UInt32 level = 0)
        {
            if (node == null)
            {
                node = _root;
            }

            Console.Write("");

            for (int i = 0; i < level * 4; i++)
            {
                Console.Write("-");
            }

            Console.Write(">");

            Console.Write(string.IsNullOrEmpty(node.Name) ? "Unnamed" : node.Name);
            Console.WriteLine($" ({node.Address})");

            foreach (var child in node.Children)
            {
                TraceNetworkTree(child, level + 1);
            }
        }

        public int NodeTimeout { get; set; }
    }
}

File: src/cpp/MasterNode.cpp
// Copyright (c) VitalElement. All rights reserved.
// Licensed under the MIT license. See licence.md file in the project root for
// full license information.
#include "MasterNode.h"
#include "Application.h"
#include "Trace.h"

constexpr uint16_t MasterNodeAddress = 1;

MasterNode::MasterNode ()
    : IdpNode (MasterGuid, "Network.Master", MasterNodeAddress)
{
    _nodesChanged = false;
    _isEnumerating = false;
    _nextAddress = 2;
    _nodeTimeout = 5000;
    _root = new NodeInfo (nullptr, MasterNodeAddress);
    _root->Guid = _guid;
    _root->Name = "Network.Master";
    _root->EnumerationState = NodeEnumerationState::Pending;

    _currentEnumerationNode = nullptr;

    _nodeInfo[Address ()] = _root;

    Manager ().RegisterCommand (
        static_cast<uint16_t> (NodeCommand::Ping),
        [&](std::shared_ptr<IncomingTransaction> incoming,
            std::shared_ptr<OutgoingTransaction> outgoing) {
            if (this->HandlePollResponse (incoming->Source ()))
            {
                return IdpResponseCode::OK;
            }
            else
            {
                return IdpResponseCode::NotReady;
            }
        });

    Manager ().RegisterCommand (
        static_cast<uint16_t> (NodeCommand::RecommendEnumeration),
        [&](std::shared_ptr<IncomingTransaction> incoming,
            std::shared_ptr<OutgoingTransaction> outgoing) {
            if (!this->IsEnumerating ())
            {
                this->EnumerateNetwork ();
            }

            return IdpResponseCode::OK;
        });

    _pollTimer = new DispatcherTimer (1000);

    _pollTimer->Tick += [&](auto sender, auto& e) {
        _pollTimer->Stop ();

        this->EnumerateNetwork ();
    };
}

MasterNode::~MasterNode ()
{
}

NodeInfo* MasterNode::NetworkTree ()
{
    return _root;
}

void MasterNode::VisitNodes (NodeInfo* root,
                             std::function<bool(NodeInfo&)> visitor)
{
    if (visitor (*root))
    {
        auto it = root->Children.begin ();

        while (it != root->Children.end ())
        {
            VisitNodes (*it, visitor);
            it++;
        }
    }
}

NodeInfo* MasterNode::GetNextEnumerationNode ()
{
    NodeInfo* result = nullptr;

    VisitNodes (_root, [&](NodeInfo& node) {
        if (node.EnumerationState != NodeEnumerationState::Idle)
        {
            result = &node;
            return false;
        }

        return true;
    });

    return result;
}

bool MasterNode::HandlePollResponse (uint16_t address)
{
    if (address != Address ())
    {
        auto node = FindNode (address);

        if (node != nullptr)
        {
            node->LastSeen = Application::GetApplicationTime ();

            return true;
        }
    }

    return false;
}

NodeInfo* MasterNode::FindNode (uint16_t address)
{
    auto it = _nodeInfo.find (address);

    if (it != _nodeInfo.end ())
    {
        return _nodeInfo[address];
    }

    return nullptr;
}

NodeInfo& MasterNode::GetNodeInfo (uint16_t address)
{
    return *_nodeInfo[address];
}

bool MasterNode::HasNode (uint16_t address)
{
    return _nodeInfo.find (address) != _nodeInfo.end ();
}

uint16_t MasterNode::GetFreeAddress ()
{
    if (_freeAddresses.empty ())
    {
        return _nextAddress++;
    }
    else
    {
        auto result = _freeAddresses.top ();

        _freeAddresses.pop ();

        return result;
    }
}

bool MasterNode::IsEnumerating ()
{
    return _isEnumerating;
}

void MasterNode::OnEnumerate ()
{
    auto node = GetNextEnumerationNode ();

    if (node != nullptr)
    {
        _currentEnumerationNode = node;

        if (node == _root)
        {
            _currentEnumerationNode->EnumerationState =
                NodeEnumerationState::DetectingRouter;

            DetectRouter ();
        }
        else if (node->IsRouter ())
        {
            switch (node->EnumerationState)
            {
                case NodeEnumerationState::Pending:
                case NodeEnumerationState::EnumeratingNodes:
                    _currentEnumerationNode->EnumerationState =
                        NodeEnumerationState::EnumeratingNodes;
                    EnumerateRouterNode (node->Address);
                    break;

                case NodeEnumerationState::StartEnumeratingAdaptors:
                    StartEnumerateRouterAdaptors (node->Address);
                    break;

                case NodeEnumerationState::EnumeratingAdaptors:
                    EnumerateRouterAdaptor (node->Address);
                    break;

                default:
                    break;
            }
        }
    }
    else if (_currentEnumerationNode != nullptr)
    {
        _currentEnumerationNode = nullptr;

        _isEnumerating = false;

        PollNetwork ();

        if (_nodesChanged)
        {
            _nodesChanged = false;
            Trace::WriteLine ("Network structure changed", "MasterNode");
            TraceNetworkTree ();
        }

        _pollTimer->Start ();
    }
}

void MasterNode::ResetNetwork ()
{
    SendRequest (0, OutgoingTransaction::Create (
                        static_cast<uint16_t> (NodeCommand::Reset),
                        CreateTransactionId (), IdpCommandFlags::None));
}

void MasterNode::OnReset ()
{
}

void MasterNode::EnumerateNetwork ()
{
    if (Connected () && !_isEnumerating)
    {
        _isEnumerating = true;

        VisitNodes (_root, [&](NodeInfo& node) {
            if (node.IsRouter ())
            {
                node.EnumerationState = NodeEnumerationState::Pending;
            }
            else
            {
                if (&node == _root)
                {
                    if (node.Children.size () == 0)
                    {
                        node.EnumerationState = NodeEnumerationState::Pending;
                    }
                }
                else
                {
                    node.EnumerationState = NodeEnumerationState::Idle;
                }
            }

            return true;
        });

        OnEnumerate ();
    }
}

void MasterNode::DetectRouter ()
{
    auto address = GetFreeAddress ();

    auto outgoingTransaction = OutgoingTransaction::Create (
        static_cast<uint16_t> (NodeCommand::RouterDetect),
        CreateTransactionId ());

    outgoingTransaction->Write (address);

    if (!SendRequest (0xFFFF, outgoingTransaction,
                      [&, address](std::shared_ptr<IdpResponse> response) {
                          if (response != nullptr &&
                              response->ResponseCode () == IdpResponseCode::OK)
                          {
                              auto nodeEnumerated =
                                  response->Transaction ()->Read<bool> ();

                              if (nodeEnumerated)
                              {
                                  _currentEnumerationNode->EnumerationState =
                                      NodeEnumerationState::Idle;

                                  this->OnNodeAdded (this->Address (), address);

                                  return;
                              }
                          }

                          _freeAddresses.push (address);

                          this->OnEnumerate ();
                      }))
    {
        _currentEnumerationNode->EnumerationState = NodeEnumerationState::Idle;

        this->OnEnumerate ();
    }
}

void MasterNode::EnumerateRouterNode (uint16_t routerAddress)
{
    auto address = GetFreeAddress ();

    auto outgoingTransaction =
        OutgoingTransaction::Create (
            static_cast<uint16_t> (NodeCommand::RouterEnumerateNode),
            CreateTransactionId ())
            ->Write (address);

    bool sent = SendRequest (
        routerAddress, outgoingTransaction,
        [&, address, routerAddress](std::shared_ptr<IdpResponse> response) {
            if (response != nullptr &&
                response->ResponseCode () == IdpResponseCode::OK)
            {
                auto nodeEnumerated = response->Transaction ()->Read<bool> ();

                if (nodeEnumerated)
                {
                    this->OnNodeAdded (routerAddress, address);

                    return;
                }
            }

            if (_currentEnumerationNode != nullptr)
            {
                _currentEnumerationNode->EnumerationState =
                    NodeEnumerationState::StartEnumeratingAdaptors;
            }

            _freeAddresses.push (address);

            this->OnEnumerate ();
        });

    if (!sent)
    {
        Trace::WriteLine ("Enumerate nodes failed.");

        _currentEnumerationNode->EnumerationState = NodeEnumerationState::Idle;

        this->OnEnumerate ();
    }
}

void MasterNode::StartEnumerateRouterAdaptors (uint16_t routerAddress)
{
    auto outgoingTransaction = OutgoingTransaction::Create (
        static_cast<uint16_t> (NodeCommand::RouterPrepareToEnumerateAdaptors),
        CreateTransactionId ());

    bool sent =
        SendRequest (routerAddress, outgoingTransaction,
                     [&](std::shared_ptr<IdpResponse> response) {
                         if (response != nullptr &&
                             response->ResponseCode () == IdpResponseCode::OK)
                         {
                             if (_currentEnumerationNode != nullptr)
                             {
                                 _currentEnumerationNode->EnumerationState =
                                     NodeEnumerationState::EnumeratingAdaptors;
                             }
                         }
                         else
                         {
                             if (_currentEnumerationNode != nullptr)
                             {
                                 _currentEnumerationNode->EnumerationState =
                                     NodeEnumerationState::Idle;
                             }
                         }

                         this->OnEnumerate ();
                     });

    if (!sent)
    {
        Trace::WriteLine ("Begin Enumerate adaptors failed.");

        _currentEnumerationNode->EnumerationState = NodeEnumerationState::Idle;

        this->OnEnumerate ();
    }
}

void MasterNode::EnumerateRouterAdaptor (uint16_t routerAddress)
{
    auto address = GetFreeAddress ();

    auto routerDetectTransactionId = CreateTransactionId ();

    auto outgoingTransaction =
        OutgoingTransaction ::Create (
            static_cast<uint16_t> (NodeCommand::RouterEnumerateAdaptor),
            CreateTransactionId ())
            ->Write (address)
            ->Write (routerDetectTransactionId);

    Manager ().RegisterOneTimeResponseHandler (
        routerDetectTransactionId,
        [&, address, routerAddress,
         routerDetectTransactionId](std::shared_ptr<IdpResponse> response) {
            if (response != nullptr &&
                response->ResponseCode () == IdpResponseCode::OK)
            {
                auto nodeEnumerated = response->Transaction ()->Read<bool> ();

                if (nodeEnumerated)
                {
                    this->OnNodeAdded (routerAddress, address);

                    auto markAdaptorConnected = OutgoingTransaction::Create (
                        static_cast<uint16_t> (
                            NodeCommand::MarkAdaptorConnected),
                        CreateTransactionId (), IdpCommandFlags::None);

                    this->SendRequest (address, markAdaptorConnected);
                    return;
                }
                else
                {
                    if (_currentEnumerationNode != nullptr)
                    {
                        _currentEnumerationNode->EnumerationState =
                            NodeEnumerationState::Idle;
                    }
                }
            }

            _freeAddresses.push (address);

            this->OnEnumerate ();
        });


    bool sent = SendRequest (
        routerAddress, outgoingTransaction,
        [&, address, routerAddress,
         routerDetectTransactionId](std::shared_ptr<IdpResponse> response) {
            if (response != nullptr &&
                response->ResponseCode () == IdpResponseCode::OK)
            {
                auto adaptorEnumerated =
                    response->Transaction ()->Read<bool> ();

                if (adaptorEnumerated)
                {
                    auto adaptorProbed =
                        response->Transaction ()->Read<bool> ();

                    if (adaptorProbed)
                    {
                        // we need to await the detect router response
                        // from the new router.

                        // this may also timeout. reset timeout here?
                        // or set a trasaction id that gets relayed,
                        // and make timeouts work on tid not command
                        // id.
                        return;
                    }
                    else
                    {
                        this->Manager ().UnregisterOneTimeResponseHandler (
                            routerDetectTransactionId);
                    }
                }
                else
                {
                    this->Manager ().UnregisterOneTimeResponseHandler (
                        routerDetectTransactionId);

                    if (_currentEnumerationNode != nullptr)
                    {
                        _currentEnumerationNode->EnumerationState =
                            NodeEnumerationState::Idle;
                    }
                }
            }
            else
            {
                this->Manager ().UnregisterOneTimeResponseHandler (
                    routerDetectTransactionId);

                if (_currentEnumerationNode != nullptr)
                {
                    _currentEnumerationNode->EnumerationState =
                        NodeEnumerationState::Idle;
                }
            }

            _freeAddresses.push (address);
            this->OnEnumerate ();
        });

    if (!sent)
    {
        this->Manager ().UnregisterOneTimeResponseHandler (
            routerDetectTransactionId);

        Trace::WriteLine ("Enumerate adaptor failed.");

        _currentEnumerationNode->EnumerationState = NodeEnumerationState::Idle;

        this->OnEnumerate ();
    }
}

void MasterNode::OnNodeAdded (uint16_t parentAddress, uint16_t address)
{
    _nodesChanged = true;

    _nodeInfo[address] = new NodeInfo (_nodeInfo[parentAddress], address);

    _nodeInfo[parentAddress]->Children.push_back (_nodeInfo[address]);

    auto nodeInfoTransaction = OutgoingTransaction::Create (
        static_cast<uint16_t> (NodeCommand::GetNodeInfo),
        CreateTransactionId ());

    SendRequest (address, nodeInfoTransaction,
                 [&, address](std::shared_ptr<IdpResponse> response) {
                     if (response != nullptr &&
                         response->ResponseCode () == IdpResponseCode::OK)
                     {
                         _nodeInfo[address]->Guid =
                             response->Transaction ()->ReadGuid ();

                         if (_nodeInfo[address]->Name != nullptr)
                         {
                             delete[] _nodeInfo[address]->Name;
                         }

                         _nodeInfo[address]->Name =
                             response->Transaction ()->ReadCString ();

                         if (_nodeInfo[address]->IsRouter ())
                         {
                             _nodeInfo[address]->EnumerationState =
                                 NodeEnumerationState::Pending;
                         }
                         else
                         {
                             _nodeInfo[address]->EnumerationState =
                                 NodeEnumerationState::Idle;
                         }

                         this->OnEnumerate ();
                     }
                 });
}

void MasterNode::InvalidateNodes ()
{
    auto currentTime = Application::GetApplicationTime ();

    auto it = _nodeInfo.begin ();

    while (it != _nodeInfo.end ())
    {
        auto address = it->first;
        auto current = it->second;

        if (current != _root && currentTime >= current->LastSeen + _nodeTimeout)
        {
            auto childIt = current->Children.begin ();

            while (childIt != current->Children.end ())
            {
                (*childIt)->Parent = nullptr;

                childIt++;
            }

            if (current->Parent != nullptr)
            {
                current->Parent->Children.remove (current);
            }

            it = _nodeInfo.erase (it);

            delete current;
            _freeAddresses.push (address);

            _nodesChanged = true;
        }
        else
        {
            it++;
        }
    }
}

void MasterNode::PollNetwork ()
{
    InvalidateNodes ();
}

uint32_t MasterNode::NodeTimeout ()
{
    return _nodeTimeout;
}

void MasterNode::NodeTimeout (uint32_t value)
{
    _nodeTimeout = value;
}

void MasterNode::TraceNetworkTree (NodeInfo* node, uint32_t level)
{
    if (node == nullptr)
    {
        node = _root;
    }

    Trace::Write ("");

    for (uint32_t i = 0; i < level * 4; i++)
    {
        Trace::Append ("-");
    }

    Trace::Append (">");

    Trace::Append (node->Name == nullptr ? "Unnamed" : node->Name);
    Trace::AppendLine (" (%u)", node->Address);

    auto it = node->Children.begin ();

    while (it != node->Children.end ())
    {
        TraceNetworkTree (*it, level + 1);
        it++;
    }
}
